### PR TITLE
Zadaca3

### PR DESCRIPTION
**Pull request by LoraLamia**

### CODE REVIEW

File: MovieApp/AppRouter.swift [hunk 1]
```diff
@@ -0,0 +1,42 @@
+
+import Foundation
+import UIKit
+import MovieAppData
+
+class AppRouter {
+    
+    private let navigationController: UINavigationController!
```
Comment (istaudacher) [inline MovieApp/AppRouter.swift]: This can be just `private let navigationController: UINavigationController!`

File: MovieApp/AppRouter.swift [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            navigationController,
+            favoritesViewController
+        ]
+        window?.rootViewController = tabBarController
+        window?.makeKeyAndVisible()
+    }
+    
+    func showMovie(movieDetails: MovieDetailsModel) {
+        let movieDetailsViewController = MovieDetailsViewController(movieDetailsModel: movieDetails)
+        self.navigationController?.pushViewController(movieDetailsViewController, animated: true)
```
Comment (istaudacher) [inline MovieApp/AppRouter.swift]: Why not just `navigationController.pushViewController(movieDetailsViewController, animated: true)`?

File: MovieApp/MovieCategoriesViewController.swift [hunk 3]
```diff
@@ -9,12 +9,22 @@ class MovieCategoriesViewController: UIViewController, UITableViewDataSource {
     private var categoryTitles = ["What's popular", "Free to watch", "Trending"]
     private var categoryMovies = [MovieUseCase().popularMovies, MovieUseCase().freeToWatchMovies, MovieUseCase().trendingMovies]
     private var categoriesTableView: UITableView!
+    var router: AppRouter!
```
Comment (istaudacher) [inline MovieApp/MovieCategoriesViewController.swift]: This should be private.

File: MovieApp/AppRouter.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        self.navigationController = navigationController
+    }
+    
+    func setStartScreen(in window: UIWindow?) {
+        let favoritesViewController = FavoritesViewController()
+        favoritesViewController.tabBarItem = UITabBarItem(
+            title: "Favorites",
+            image: UIImage(named: "tabFavorites"),
+            selectedImage: UIImage(named: "tabFavoritesSelected"))
+        let movieCategoriesViewController = MovieListViewController(router: self)
```
Comment (istaudacher) [inline MovieApp/AppRouter.swift]: This should be `MovieCategoriesViewController(router: self)`

File: MovieApp/MovieListViewController.swift [hunk 5]
```diff
@@ -7,6 +7,7 @@ import Kingfisher
 class MovieListViewController: UIViewController, UITableViewDelegate {
     
     private var movieListTableView: UITableView!
+    var router: AppRouter!
```
Comment (istaudacher) [inline MovieApp/MovieListViewController.swift]: This should be private.

File: MovieApp/MovieListViewController.swift [hunk 6]
```diff
@@ -62,7 +73,11 @@ extension MovieListViewController: UITableViewDataSource {
         } else {
             return UITableViewCell()
         }
-        
     }
-
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let movies = MovieUseCase().allMovies
```
Comment (istaudacher) [inline MovieApp/MovieListViewController.swift]: You should fetch all movies on viewDidLoad and store them in a local field. This is an expensive call.

File: MovieApp/MoviesTableViewCell.swift [hunk 7]
```diff
@@ -4,11 +4,16 @@ import PureLayout
...
+protocol MovieCollectionCellDelegate {
+    func didSelectMovie(movieDetails: MovieDetailsModel)
+}
+
+class MoviesTableViewCell: UITableViewCell, UICollectionViewDataSource, UICollectionViewDelegate {
     
     static let identifier = "CollectionTableViewCell"
     private var categoryLabel: UILabel!
     private var collectionView: UICollectionView!
+    var delegate: MovieCollectionCellDelegate?
```
Comment (istaudacher) [inline MovieApp/MoviesTableViewCell.swift]: `private weak var delegate: MovieCollectionCellDelegate?`

File: MovieApp/AppRouter.swift [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+class AppRouter {
+    
+    private let navigationController: UINavigationController!
+    
+    init(navigationController: UINavigationController) {
+        self.navigationController = navigationController
+    }
+    
+    func setStartScreen(in window: UIWindow?) {
```
Comment (istaudacher) [inline MovieApp/AppRouter.swift:16]: Although you have achieved the correct behavior, it is not implemented in the correct way. What you did is:
* You have root NavigationController, passed in the constructor
* You have pushed MovieViewController in the root NavigationController
* You have assigned root NavigationController and FavoritesVC in the TabVC

The correct way would be:
* In the constructor, initialize two additional NavigationControllers and assign them to a class field
* One NC is for the Movie tab, other one is for the Favorites tab
* Push MovieVC to the first NC and push FavoritesVC to the second NC
* Assign those two NC to the TabVC
* In `showMovie`, use movie NC to push `MovieDetailsVC`

File: MovieApp/AppRouter.swift (hunk 1 above)
Comment (LoraLamia) [reply MovieApp/AppRouter.swift]: Not sure what do you mean? 😅

**Pull request merged**